tests/processor/test_pipeline.py

```diff
@@ -2146,3 +2195,149 @@ def custom_to_output(transition):
         # Should work with standard format (wouldn't work with custom converter)
         result = loaded(batch)
         assert "observation.image" in result  # Standard format preserved
+
+
+class NonCompliantStep:
+    """Intentionally non-compliant: missing feature_contract."""
+
+    def __call__(self, transition: EnvTransition) -> EnvTransition:
+        return transition
+
+
+def test_construction_rejects_step_without_feature_contract():
+    with pytest.raises(TypeError, match=r"must define feature_contract\(features\) -> dict\[str, Any\]"):
+        RobotProcessor([NonCompliantStep()])
+
+
+class NonCallableStep:
+    """Intentionally non-compliant: missing __call__."""
+
+    def feature_contract(self, features: dict[str, Any]) -> dict[str, Any]:
+        return features
+
+
+def test_construction_rejects_step_without_call():
+    with pytest.raises(TypeError, match=r"must define __call__"):
+        RobotProcessor([NonCallableStep()])
+
+
+@dataclass
+class FeatureContractAddStep:
+    """Adds a key to the features dict"""
+
+    key: str = "a"
+    value: Any = 1
+
+    def __call__(self, transition: EnvTransition) -> EnvTransition:
+        return transition
+
+    def feature_contract(self, features: dict[str, Any]) -> dict[str, Any]:
+        features[self.key] = self.value
+        return features
+
+
+@dataclass
+class FeatureContractMutateStep:
+    key: str = "a"
+    fn: Callable[[Any], Any] = lambda x: x  # noqa: E731
+
+    def __call__(self, transition: EnvTransition) -> EnvTransition:
+        return transition
+
+    def feature_contract(self, features: dict[str, Any]) -> dict[str, Any]:
+        features[self.key] = self.fn(features.get(self.key))
+        return features
+
+
+@dataclass
+class FeatureContractBadReturnStep:
+    def __call__(self, transition: EnvTransition) -> EnvTransition:
+        return transition
+
+    def feature_contract(self, features: dict[str, Any]):
+        return ["not-a-dict"]
+
+
+@dataclass
+class FeatureContractRemoveStep:
+    key: str
+
+    def __call__(self, transition: EnvTransition) -> EnvTransition:
+        return transition
+
+    def feature_contract(self, features: dict[str, Any]) -> dict[str, Any]:
+        features.pop(self.key, None)
+        return features
+
+
+def test_feature_contract_empty_pipeline_returns_deepcopy():
+    p = RobotProcessor()
+    out0 = p.feature_contract()
+    assert out0 == {}
```

Comment:
this case should be removed if we remove `None` params in `RobotProcessor.feature_contract`